crosscluster, jobs: create reader tenant job

azhu-crl opened the pull request:
This commit creates a new job type `StandbyReadTSPollerJob`. `StandbyReadTSPollerJob` polls the `ReadFromTenantInfo` API every 15s and updates replicated time for the reader tenant if it has advanced.

Epic: CRDB-23575
Informs: https://github.com/cockroachdb/cockroach/issues/130445
Release note: None

azhu-crl committed:
crosscluster, jobs: create reader tenant job

This commit creates a new job type `StandbyReadTSPollerJob`.
`StandbyReadTSPollerJob` polls the `ReadFromTenantInfo` API
every 15s and updates replicated time for the reader tenant
if it has advanced.

Epic: CRDB-23575
Informs: https://github.com/cockroachdb/cockroach/issues/130445
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130696)
<!-- Reviewable:end -->

dt commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job_test.go:
```diff
@@ -0,0 +1,44 @@
+	c.WaitUntilReplicatedTime(srcTime, jobspb.JobID(ingestionJobID))
+
+	// TODOs:
+	// 1) create job record
```
I think this should be pretty easy; you can get the jobRegistry from the testserver

azhu-crl replied:
Got it thanks! I added testing for verifying job creation/status, as for verifying correctness of replicated timestamp of the reader tenant, should we add that in the bootstrap migration pr so that we don't have to repeat the logic of calling `SetupOrAdvanceStandbyReaderCatalog` in this test? I added a TODO here as well

dt replied:
I would probably lean towards testing this e2e by inserting a row into the src tenant then reading it back in a succeedsSoon from the reader.

msbutler commented:
@azhu-crl a heads up that CI found null pointer bug in your code. I'm going to wait to review until you address that.

dt commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job_test.go:
```diff
@@ -0,0 +1,73 @@
+	// TODO(azhu): verify replicated timestamp of reader tenant has advanced
+}
+
+func makeStandbyReadTSPollerJobRecord(registry *jobs.Registry) jobs.Record {
```
I'd move this to the non-_test file, and export it, since the bootstrap job will want to call this, right?

dt commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job.go:
```diff
@@ -0,0 +1,100 @@
+type standbyReadTSPollerResumer struct {
+	job *jobs.Job
+
+	readerTenant mtinfo.ReadFromTenantInfoAccessor
```
can we plumb this via the `sql.JobExecContext` ?

msbutler commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job.go:
```diff
@@ -0,0 +1,123 @@
+			return ctx.Err()
+		case <-ticker.C:
+			tenantInfoAccessor := execCfg.UpgradeJobDeps.SystemDeps().TenantInfoAccessor
+			if tenantInfoAccessor == nil {
```
i think this should return an assertionFailed error. It's ok to internally fail a non cancellable job if the failure is justified. We can't recover from a nil `TenantInfoAccessor`

msbutler commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job.go:
```diff
@@ -0,0 +1,123 @@
+		err := errors.NewAssertionErrorWithWrappedErrf(
+			jobErr, "standby read ts poller job is not cancelable",
+		)
+		log.Errorf(ctx, "%v", err)
```
we should throw an `errors.AssertionFailedf` here if the job was cancelled. It would suggest that something is deeply wrong with the job system

msbutler reviewed (commented):
nearly there!

azhu-crl committed:
server, upgrade: plumb `TenantInfoAccessor` field

Add a `TenantInfoAccessor` field in `SystemDeps` such that
we can access reader tenant in `StandbyReadTSPollerJob`.

Epic: CRDB-23575
Informs: https://github.com/cockroachdb/cockroach/issues/130445
Release note: None

msbutler commented on pkg/ccl/crosscluster/physical/standby_read_ts_poller_job_test.go:
```diff
@@ -0,0 +1,95 @@
+func TestStandbyReadTSPollerJob(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
```
is this test still taking a long time? could you speed it up by setting the cluster setting on the poll frequency to 100 milliseconds?

azhu-crl replied:
It completes under 1 minute as opposed to 45s now when polling at 1 microsecond interval (which is less than 100 milliseconds), I think it's probably the best we can do

msbutler replied:
you. mean under 1 second?

msbutler reviewed (approved):
LGTM module the test completes in a couple seconds.

azhu-crl commented:
TFTR!

bors r=msbutler
